specs/core/1_custody-game.md

```diff
@@ -0,0 +1,483 @@
+
+#### Custody reveals
+
+Verify that `len(block.body.custody_reveals) <= MAX_CUSTODY_REVEALS`.
+
+For each `reveal` in `block.body.custody_reveals`, run the following function:
+
+```python
+def process_custody_reveal(state: BeaconState,
+                           reveal: CustodyReveal) -> None:
+    assert verify_custody_reveal(reveal)
+    revealer = state.validator_registry[reveal.revealer_index]
+
+    # Case 1: non-masked non-punitive non-early reveal
+    if reveal.mask == ZERO_HASH:
+        assert reveal.period == epoch_to_custody_period(revealer.activation_epoch) + revealer.custody_reveal_index
+        # Revealer is active or exited
+        assert is_active_validator(revealer, get_current_epoch(state)) or revealer.exit_epoch > get_current_epoch(state)
+        revealer.custody_reveal_index += 1
+        revealer.max_reveal_lateness = max(revealer.max_reveal_lateness, get_current_period(state) - reveal.period)
+
+    # Case 2: masked punitive early reveal
+    else:
+        assert reveal.period > epoch_to_custody_period(get_current_epoch(state))
+        assert revealer.slashed is False
+        slash_validator(state, reveal.revealer_index, reveal.masker_index)
+        increase_balance(state, reveal.masker_index, base_reward(state, index) // MINOR_REWARD_QUOTIENT)
+
+    proposer_index = get_beacon_proposer_index(state, state.slot)
+    increase_balance(state, proposer_index, base_reward(state, index) // MINOR_REWARD_QUOTIENT)
+
+```
+
+#### Chunk challenges
+
+Verify that `len(block.body.chunk_challenges) <= MAX_CHUNK_CHALLENGES`.
+
+For each `challenge` in `block.body.chunk_challenges`, run the following function:
+
+```python
+def process_chunk_challenge(state: BeaconState,
+                            challenge: ChunkChallenge) -> None:
+    # Verify the attestation
+    assert verify_standalone_attestation(state, convert_to_standalone(state, challenge.attestation))
+    # Verify it is not too late to challenge
+    assert slot_to_epoch(challenge.attestation.data.slot) >= get_current_epoch(state) - MAX_CHUNK_CHALLENGE_DELAY
+    assert state.validator_registry[responder_index].exit_epoch >= get_current_epoch(state) - MAX_CHUNK_CHALLENGE_DELAY
+    # Verify the responder participated in the attestation
+    assert challenger.responder_index in challenge.attestation.validator_indices
+    # Verify the challenge is not a duplicate
+    for record in state.chunk_challenge_records:
+        assert (
+            record.crosslink_data_root != challenge.attestation.data.crosslink_data_root or
+            record.chunk_index != challenge.chunk_index
+        )
+    # Verify depth
+    depth = math.log2(next_power_of_two(get_crosslink_chunk_count(challenge.attestation)))
+    assert challenge.chunk_index < 2**depth
+    # Add new chunk challenge record
+    state.chunk_challenge_records.append(ChunkChallengeRecord(
+        challenge_index=state.challenge_index,
+        challenger_index=get_beacon_proposer_index(state, state.slot),
```

Comment:
I am not a fan of forcing the chunk challenger to be the proposer, but I agree it is much simpler.

Reply:
This is also a question I have. To be discussed on today's call.